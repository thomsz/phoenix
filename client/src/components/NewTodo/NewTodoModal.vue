<template>
  <div class="flex flex-col space-y-3 absolute w-full bg-white/80 backdrop-blur-sm top-16">
    <hr />
    <textarea
      name="description"
      rows="5"
      class="w-full p-2 bg-transparent border-slate-100 border-b-2"
      v-model="form.description"
      placeholder="Descriptionâ€¦"
    >
    </textarea>
    <label class="px-2 text-gray-400 hover:cursor-pointer">
      Due date:
      <input
        name="due_date"
        type="date"
        class="h-8 bg-transparent"
        v-model="form.dueDate"
      />
    </label>
    <div class="w-full flex gap-2">
      <button
        class="action bg-blue-100 hover:bg-blue-200 grow"
        :class="{ 'bg-blue-400 hover:bg-blue-500': form.title }"
      >
        Create task
      </button>
      <button
        type="button"
        class="action w-12 hover:bg-gray-300"
        @click="$emit('close-button-clicked')"
      >
        <XIcon class="close icon fill-slate-300" />
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import { XIcon } from '@heroicons/vue/solid'

export default defineComponent({
  name: 'NewTodoModal',

  components: {
    XIcon
  },

  props: {
    modelValue: {
      type: Object,
      required: true
    }
  },

  data () {
    return {
      form: this.modelValue
    }
  },

  watch: {
    form: {
      deep: true,
      handler (form): void {
        this.$emit('update:modelValue', form)
      }
    }
  }
})
</script>

<style scoped>
button.action {
  @apply flex justify-center items-center h-12 px-3 rounded-md font-bold uppercase text-white transition-colors
    duration-500;
}

button.action:hover .icon.close {
  @apply fill-white;
}
</style>
