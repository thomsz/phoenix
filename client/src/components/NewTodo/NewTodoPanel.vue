<template>
  <form
    ref="newTodo"
    name="new_todo"
    @submit.prevent="createTodo"
  >
    <input
      name="title"
      v-model="form.title"
      required
    />
    <div>
      <input
        name="description"
        v-model="form.description"
      />
      <input
        name="due_date"
        type="date"
        v-model="form.dueDate"
      />
      <button>Add</button>
    </div>
  </form>
</template>

<script lang="ts">
import { v4 as uuid } from 'uuid'
import { defineComponent } from 'vue'

import type Todo from '@/types/Todo'

export default defineComponent({
  name: 'NewTodoPanel',

  data: () => ({
    form: {
      title: '',
      dueDate: '',
      description: ''
    }
  }),

  methods: {
    resetForm (): void {
      // TODO: typing
      (this.$refs.newTodo as any).reset()
    },

    createTodo (): void {
      // TODO: create todo
      const { title, dueDate, description } = this.form
      const todo: Todo = {
        id: uuid(),
        title,
        description,
        ...(dueDate && { dueDate: new Date(dueDate) }),
      }

      this.resetForm()
      this.$emit('todo-created', todo)
    }
  }
})
</script>
