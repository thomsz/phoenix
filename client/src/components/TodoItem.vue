<template>
  <article
    class="hover:cursor-pointer hover:bg-slate-50 transition-colors duration-300"
    :class="{ 'opacity-25': item.done }"
  >
    <div class="flex justify-between items-center h-12">
      <input
        type="checkbox"
        class="mx-2 h-full w-7 accent-green-600 hover:cursor-pointer"
        v-model="form.done"
      />
      <EditInPlaceField
        name="title"
        v-model="form.title"
      />
      <button
        class="action"
        @click="expanded = !expanded"
      >
        Details
      </button>
      <button
        class="action"
        @click="$emit('delete-item-clicked')"
      >
        Delete
      </button>
    </div>
    <div
      v-if="expanded"
      class="p-4"
    >
      <EditInPlaceField
        name="description"
        v-model="form.description"
      />
      <EditInPlaceField
        name="due_date"
        type="date"
        v-model="form.dueDate"
      />
    </div>
  </article>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import EditInPlaceField from '@/components/EditInPlaceField.vue'

import type { PropType } from 'vue'
import type Todo from '@/types/Todo'

export default defineComponent({
  name: 'TodoItem',

  components: {
    EditInPlaceField
  },

  props: {
    item: {
      type: Object as PropType<Todo>,
      required: true
    }
  },

  data () {
    return {
      form: this.item,
      expanded: false
    }
  },

  methods: {
    onItemChange (): void {
      this.$emit('item-changed', this.form)
    }
  }
})
</script>

<style scoped>
button.action {
  @apply h-7 px-2 rounded bg-blue-50 border-2 border-blue-200 m-2 text-blue-900 font-bold;
}
</style>
