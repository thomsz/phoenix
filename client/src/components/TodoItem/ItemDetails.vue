<template>
  <div class="p-4">
    <EditInPlace
      name="description"
      type="textarea"
      class="min-h-[2.5rem]"
      v-model="form.description"
    >
      <span v-if="form.description">{{ form.description }}</span>
      <span
        v-else
        class="text-slate-200"
      >
        Add description
      </span>
    </EditInPlace>
    <hr />
    <EditInPlace
      name="due_date"
      type="date"
      class="min-h-[2.5rem]"
      v-model="form.dueDate"
    >
      <span v-if="form.dueDate">{{ formattedLocaleDueDate }}</span>
      <span
        v-else
        class="text-slate-200"
      >
        Add due date
      </span>
    </EditInPlace>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import EditInPlace from '@/components/common/EditInPlace.vue'

import type Todo from '@/types/Todo'
import type { PropType } from 'vue'

export default defineComponent({
  name: 'ItemDetails',

  components: {
    EditInPlace
  },

  props: {
    modelValue: {
      type: Object as PropType<Todo>,
      required: true
    }
  },

  data () {
    return {
      form: this.modelValue
    }
  },

  computed: {
    formattedLocaleDueDate (): string | undefined {
      return new Date(this.form.dueDate).toLocaleDateString()
    }
  },

  watch: {
    form: {
      deep: true,
      handler (form: Todo) {
        this.$emit('update:modelValue', form)
      }
    }
  }
})
</script>
