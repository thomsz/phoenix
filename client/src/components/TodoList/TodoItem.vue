<template>
  <div>
    {{ item.title }}
    {{ item.description || '' }}
    {{ item.dueDate || '' }}
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import type { PropType } from 'vue'
import type Todo from '@/types/Todo'

export default defineComponent({
  name: 'TodoItem',

  props: {
    item: {
      type: Object as PropType<Todo>,
      required: true
    }
  }
})
</script>
