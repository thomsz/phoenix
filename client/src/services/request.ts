import env from '@/services/env'
import request from 'axios'
import type { AxiosInstance } from 'axios'

const generateRequestInstance = (): AxiosInstance => request.create({
  baseURL: env.API_URL
})

export default generateRequestInstance()
