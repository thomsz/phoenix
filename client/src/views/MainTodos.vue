<template>
  <NewTodoPanel @todo-created="addItem" />
  <TodoItem
    v-for="item in todos"
    :key="item.id"
    :item="item"
    @item-changed="updateItem"
    @delete-item-clicked="deleteItem(item)"
  />
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import TodoItem from '@/components/TodoItem.vue'
import NewTodoPanel from '@/components/NewTodo/NewTodoPanel.vue'

import type Todo from '@/types/Todo'

export default defineComponent({
  name: 'MainTodos',

  components: {
    TodoItem,
    NewTodoPanel
  },

  data: () => ({
    todos: [] as Array<Todo>
  }),

  mounted (): void {
    this.fetchTodos()
  },

  methods: {
    addItem (item: Todo): void {
      this.todos.push(item)
    },

    updateItem (item: Todo): void {
      // TODO: update item
      const itemIndex = this.todos.findIndex((todo: Todo) => todo.id == item.id)
      this.todos[itemIndex] = item
    },

    deleteItem (item: Todo): void {
      // TODO: delete item
      const itemIndex = this.todos.findIndex((todo: Todo) => todo.id == item.id)
      this.todos.splice(itemIndex, 1)
    },

    fetchTodos (): void {
      // TODO: fetch from server
      this.todos = [
        {
          id: '1',
          title: 'Item 1',
          description: 'Item 1 description'
        },
        {
          id: '2',
          title: 'Item 2',
          description: 'Item 2 description'
        },
        {
          id: '3',
          title: 'Item 3',
          description: 'Item 3 description'
        }
      ]
    }
  }
})
</script>
