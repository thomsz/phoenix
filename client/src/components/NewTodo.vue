 <template>
  <form
    name="new_todo"
    @submit.prevent="createTodo"
  >
    <div class="flex">
      <input
        name="title"
        class="w-full h-16 text-2xl px-2"
        v-model="form.title"
        required
        placeholder="Add a new taskâ€¦"
        @focus="expanded = true"
      />
      <ReturnIcon
        v-if="expanded"
        class="w-10 mx-2 fill-slate-200 hidden md:block"
      />
    </div>
    <Transition name="fade">
      <NewTodoModal
        v-if="expanded"
        v-model="form"
        @close-button-clicked="expanded = false"
      />
    </Transition>
  </form>
</template>

<script lang="ts">
import request from '@/services/request'
import { defineComponent } from 'vue'

import ReturnIcon from '@/components/icons/ReturnIcon.vue'
import NewTodoModal from '@/components/NewTodo/NewTodoModal.vue'

import type Todo from '@/types/Todo'

const generateForm = () => ({
  title: '',
  dueDate: '',
  description: ''
})

export default defineComponent({
  name: 'NewTodo',

  components: {
    ReturnIcon,
    NewTodoModal
  },

  data: () => ({
    expanded: false,
    form: generateForm()
  }),

  watch: {
    'form.title' (title: string): void {
      if (title) {
        this.expanded = true
      }
    }
  },

  methods: {
    resetForm (): void {
      this.expanded = false
      this.form = generateForm()
    },

    async createTodo (): Promise<void> {
      const { title, dueDate, description } = this.form
      const item: Omit<Todo, 'id'> = {
        done: false,
        title,
        dueDate: dueDate ? new Date(dueDate).toISOString() : '',
        description
      }

      try {
        const response = await request.post('tasks', item)
        const createdItem: Todo = response.data
        this.resetForm()
        this.$emit('todo-created', createdItem)
      } catch (error) {
        console.error(error)
      }

    }
  }
})
</script>

<style scoped>
form {
  @apply flex flex-col relative z-10;
}

.fade-enter-active {
  @apply duration-300 ease-out;
}

.fade-enter-from {
  @apply opacity-0;
}

.fade-enter-to {
  @apply opacity-100;
}

.fade-leave-active {
  @apply duration-200 ease-in;
}

.fade-leave-from {
  @apply opacity-100;
}

.fade-leave-to {
  @apply opacity-0;
}
</style>
