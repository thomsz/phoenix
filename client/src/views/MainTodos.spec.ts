import request from '@/services/request'
import factory from '@/tests/services/factory'
import { faker } from '@faker-js/faker'
import { shallowMount } from '@vue/test-utils'

import NewTodo from '@/components/NewTodo.vue'
import TodoItem from '@/components/TodoItem.vue'
import MainTodos from '@/views/MainTodos.vue'

import type Todo from '@/types/Todo'

jest.mock('@/services/env', () => ({}))
jest.mock('@/services/request', () => ({
  get: (url: string) => {
    switch (url) {
      case 'tasks':
        return Promise.resolve({
          data: []
        })
    
      default:
        return Promise.reject()
    }
  },
  put: jest.fn(),
  delete: jest.fn()
}))

describe('MainTodos', () => {
  [
    ['no items', []],
    ['with items', [factory.generate('Todo')]]
  ].forEach(([descriptor, tasks]) => {
    it(`should render fetched tasks: ${descriptor}`, async () => {
      jest.spyOn(request, 'get').mockImplementation((url: string) => {
        switch (url) {
          case 'tasks':
            return Promise.resolve({
              data: tasks
            })

          default:
            return Promise.reject()
        }
      })

      const wrapper = await shallowMount(MainTodos)
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('should create a new item', async () => {
    const wrapper = shallowMount(MainTodos)
    expect(wrapper.findAllComponents(TodoItem).length).toBe(0)

    const newTodoWrapper = wrapper.findComponent(NewTodo)
    await newTodoWrapper.vm.$emit('todo-created', factory.generate('Todo'))
    expect(wrapper.findAllComponents(TodoItem).length).toBe(1)
  })

  it('should update item', async () => {
    const fakeItem: Todo = factory.generate('Todo')
    const fakeUpdatedItem = {
      ...factory.generate('Todo'),
      id: fakeItem.id
    }

    jest.spyOn(request, 'put').mockResolvedValue({
      data: fakeUpdatedItem
    })

    const wrapper = shallowMount(MainTodos, {
      data: () => ({
        todos: [fakeItem]
      })
    })

    const todoItemWrapper = wrapper.findComponent(TodoItem)
    expect(todoItemWrapper.props('item')).toEqual(fakeItem)

    await todoItemWrapper.vm.$emit('item-changed', fakeUpdatedItem)
    expect(request.put).toBeCalledWith(`tasks/${fakeItem.id}`, fakeUpdatedItem)
  })

  it('should delete item', async () => {
    jest.spyOn(request, 'delete')
    const fakeItem: Todo = factory.generate('Todo')
    const wrapper = shallowMount(MainTodos, {
      data: () => ({
        todos: [fakeItem]
      })
    })

    const todoItemWrapper = wrapper.findComponent(TodoItem)
    expect(todoItemWrapper.props('item')).toEqual(fakeItem)

    await todoItemWrapper.vm.$emit('delete-item-button-clicked')
    expect(request.delete).toBeCalledWith(`tasks/${fakeItem.id}`)

    await wrapper.vm.$nextTick()
    expect(wrapper.findComponent(TodoItem).exists()).toBe(false)
  })

  it('should filter items', async () => {
    const fakeItems = Array.from({ length: 4 }, () => factory.generate('Todo'))
    const wrapper = shallowMount(MainTodos, {
      data: () => ({
        todos: fakeItems
      })
    })

    const mostRecentItem: Todo = fakeItems[fakeItems.length - 1]
    const todoItemWrappers = wrapper.findAllComponents(TodoItem)
    expect(todoItemWrappers.length).toBe(fakeItems.length)
    expect(todoItemWrappers[0].props('item')).toEqual(mostRecentItem)

    const randomItem: Todo = faker.helpers.arrayElement(fakeItems)
    const searchQueryVariations = [
      randomItem.title,
      ` ${randomItem.title} `,
      randomItem.title.toUpperCase()
    ]

    searchQueryVariations.forEach(async (searchQuery: string) => {
      await wrapper.setData({ searchQuery })
      const filteredTodoItemWrappers = wrapper.findAllComponents(TodoItem)
      expect(filteredTodoItemWrappers.length).toBe(1)
      expect(filteredTodoItemWrappers[0].props('item')).toEqual(randomItem)
    })
  })
})
