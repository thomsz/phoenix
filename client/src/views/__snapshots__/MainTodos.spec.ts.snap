// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainTodos should render fetched tasks: no items 1`] = `
"<main>
  <new-todo-stub></new-todo-stub>
  <search-bar-stub modelvalue=\\"\\"></search-bar-stub>
  <div class=\\"mx-auto py-16 text-slate-400\\"> No tasks to show </div>
</main>"
`;

exports[`MainTodos should render fetched tasks: with items 1`] = `
"<main>
  <new-todo-stub></new-todo-stub>
  <search-bar-stub modelvalue=\\"\\"></search-bar-stub>
  <div class=\\"divide-y\\">
    <todo-item-stub item=\\"[object Object]\\"></todo-item-stub>
  </div>
</main>"
`;
